File: Laboratory/TypeScript/test/union.ts
import { BebopView } from 'bebop';
import { IU, U, IWeirdOrder, WeirdOrder } from './generated/gen';
import * as assert from "assert";

it("Union roundtrip", () => {
    const obj = U.fromA({ b: 12345 });
    const bytes = U.encode(obj);
    const obj2 = U.decode(bytes);
    expect(obj).toEqual(obj2);
});

it("Union weird discriminator order roundtrip", () => {
    const obj = WeirdOrder.fromTwoComesFirst({ b: 99 });
    const bytes = WeirdOrder.encode(obj);
    const obj2 = WeirdOrder.decode(bytes);
    expect(obj).toEqual(obj2);
});

File: Laboratory/C++/test/union.cpp
#include <variant>
#include "../gen/union.hpp"

int main() {
  WeirdOrder o;

  // We can put a value in...
  o.variant = TwoComesFirst { 42 };

  // And get it back out
  TwoComesFirst x = std::get<TwoComesFirst>(o.variant);
  if (x != 42) return 1;

  return 0;
}
